From f0179279544fa45cbd05da3b33254d6969c70e20 Mon Sep 17 00:00:00 2001
From: Daniel Veillard <veillard@redhat.com>
Date: Mon, 23 Jul 2012 14:15:40 +0800
Subject: [PATCH] Do not fetch external parsed entities
To: libvir-list@redhat.com

Unless explicietely asked for when validating or replacing entities
with their value. Problem pointed out by Tom Lane <tgl@redhat.com>

* parser.c: do not load external parsed entities unless needed
* test/errors/extparsedent.xml result/errors/extparsedent.xml*:
  add a regression test to avoid change of the behaviour in the future

Signed-off-by: Daniel Veillard <veillard@redhat.com>
---
 parser.c                       | 11 +++++++++--
 result/errors/extparsedent.xml |  5 +++++
 test/errors/extparsedent.xml   |  5 +++++
 3 files changed, 19 insertions(+), 2 deletions(-)
 create mode 100644 result/errors/extparsedent.xml
 create mode 100644 result/errors/extparsedent.xml.err
 create mode 100644 result/errors/extparsedent.xml.str
 create mode 100644 test/errors/extparsedent.xml

diff --git a/parser.c b/parser.c
index 4bfc5f8..f874344 100644
--- a/parser.c
+++ b/parser.c
@@ -6789,8 +6789,15 @@ xmlParseReference(xmlParserCtxtPtr ctxt) {
      * The first reference to the entity trigger a parsing phase
      * where the ent->children is filled with the result from
      * the parsing.
-     */
-    if (ent->checked == 0) {
+     * Note: external parsed entities will not be loaded, it is not
+     * required for a non-validating parser, unless the parsing option
+     * of validating, or substituting entities were given. Doing so is
+     * far more secure as the parser will only process data coming from
+     * the document entity by default.
+     */
+    if ((ent->checked == 0) &&
+        ((ent->etype != XML_EXTERNAL_GENERAL_PARSED_ENTITY) ||
+         (ctxt->options & (XML_PARSE_NOENT | XML_PARSE_DTDVALID)))) {
 	unsigned long oldnbent = ctxt->nbentities;
 
 	/*
diff --git a/result/errors/extparsedent.xml b/result/errors/extparsedent.xml
new file mode 100644
index 0000000..07e4c54
--- /dev/null
+++ b/result/errors/extparsedent.xml
@@ -0,0 +1,5 @@
+<?xml version="1.0"?>
+<!DOCTYPE foo [
+<!ENTITY c PUBLIC "bar" "/etc/doesnotexist">
+]>
+<root>&c;</root>
diff --git a/test/errors/extparsedent.xml b/test/errors/extparsedent.xml
new file mode 100644
index 0000000..07e4c54
--- /dev/null
+++ b/test/errors/extparsedent.xml
@@ -0,0 +1,5 @@
+<?xml version="1.0"?>
+<!DOCTYPE foo [
+<!ENTITY c PUBLIC "bar" "/etc/doesnotexist">
+]>
+<root>&c;</root>
-- 
1.7.11.2

